ciphersuite/ciphersuite.go

```diff
@@ -0,0 +1,171 @@
+	return
+}
+
+// MarshalText implements the encoding interface TextMarshaler so that
+// it can be serialized in format such as TOML.
+func (d *CipherData) MarshalText() ([]byte, error) {
+	name := []byte(d.Name)
+	size := make([]byte, sizeLength)
+	binary.LittleEndian.PutUint32(size, uint32(len(name)))
+
+	// Buffer starts with the size of the cipher suite name, then the name
+	// and finally the data.
+	data := append(append(size, name...), d.Data...)
+
+	buf := make([]byte, hex.EncodedLen(len(data)))
+	hex.Encode(buf, data)
+	return buf, nil
+}
+
+// UnmarshalText implements the encoding interface TextUnmarshaler so that
+// format such as TOML can deserialize the data.
+func (d *CipherData) UnmarshalText(text []byte) error {
+	buf := make([]byte, hex.DecodedLen(len(text)))
+	_, err := hex.Decode(buf, text)
+	if err != nil {
+		return xerrors.Errorf("decoding hex: %v", err)
+	}
+
+	if len(buf) < sizeLength {
+		return xerrors.Errorf("data is too small")
+	}
+
+	size := int(binary.LittleEndian.Uint32(buf[:sizeLength]))
+	if len(buf) < sizeLength+size {
+		return xerrors.Errorf("data is too small")
+	}
+
+	d.Name = string(buf[sizeLength : sizeLength+size])
+	d.Data = buf[sizeLength+size:]
+	return nil
+}
+
+// Packable provides the primitives necessary to make network
+// messages out of interfaces.
+type Packable interface {
+	// Pack must create a cipher data container that will allow an unpacking
+	// operation to recreate the same element. This operation should not fail.
+	Pack() *CipherData
+
+	// Unpack takes a cipher data container and recreate the corresponding
+	// element. If the type does not match or an decoding error occurs, it
+	// should return the error.
+	Unpack(data *CipherData) error
+}
+
+// PublicKey represents one of the two sides of an asymmetric key pair
+// which can be safely shared publicly.
+type PublicKey interface {
+	Packable
+	Nameable
+	fmt.Stringer
+}
+
+// SecretKey represents one of the two sides of an asymmetric key pair
+// which must remain private.
+type SecretKey interface {
+	Packable
+	Nameable
+	fmt.Stringer
+}
+
+// Signature represents a signature produced using a secret key and
+// that can be verified with the associated public key.
+type Signature interface {
+	Packable
+	Nameable
+	fmt.Stringer
+}
+
+// CipherSuite provides the primitive needed to create and verify
+// signatures using an asymmetric key pair.
+type CipherSuite interface {
+	Nameable
+
+	// PublicKey must return an implementation of a public key.
+	PublicKey() PublicKey
+
+	// SecretKey must return an implementation of a secret key.
+	SecretKey() SecretKey
+
+	// Signature must return an implementation of a signature.
+	Signature() Signature
+
+	// KeyPair must return a random secret key and its associated public key.
+	KeyPair() (PublicKey, SecretKey)
+
+	// Sign must produce a signature that can be validated by the
+	// associated public key of the secret key.
+	Sign(sk SecretKey, msg []byte) (Signature, error)
```

Comment:
Not sure: is this enough for lattice-based crypto? Should there be a more generic interface?